File: LeetCode/remote/first_missing_positive.cs
// https://leetcode.com/submissions/detail/33873237/

public class Solution {
    public int FirstMissingPositive(int[] nums) {
        int i = 0;
        while (i < nums.Length) {
            while (nums[i] > 0 && nums[i] <= nums.Length && nums[i] != nums[nums[i] - 1]) {
                int aux = nums[nums[i] - 1];
                nums[nums[i] - 1] = nums[i];
                nums[i] = aux;
            }
            
            i++;
        }
        
        for (i = 0; i < nums.Length; i++) {
            if (nums[i] != i + 1) {
                return i + 1;
            }
        }
        
        return nums.Length + 1;
    }
}

//  Almost same fucking thing, only worse because of cheating
public class Solution {
    public int FirstMissingPositive(int[] nums) {
        for (var i = 0; i < nums.Length; i++)
        {
            while (nums[i] - 1 != i && 0 < nums[i] && nums[i] <= nums.Length && nums[nums[i] - 1] != nums[i])
            {
                var c = nums[nums[i] - 1];
                nums[nums[i] - 1] = nums[i];
                nums[i] = c;
            }
        }
        
        for (var i = 0; i < nums.Length; i++)
        {
            if (nums[i] != i + 1)
            {
                return i + 1;
            }
        }
        
        return nums.Length + 1;
    }
}

File: LeetCode/remote/first_missing_positive.rb
#   https://leetcode.com/problems/first-missing-positive/
#
#    Given an unsorted integer array, find the first missing positive integer. 
#
#   Well this was a great fucking struggle considering I knew how the solution should work..
#   Just to remind you how weak and small you really are.
#   
#   https://leetcode.com/submissions/detail/54023547/
#   Submission Details
#   156 / 156 test cases passed.
#       Status: Accepted
#       Runtime: 84 ms
#           
#           Submitted: 0 minutes ago

# @param {Integer[]} nums
# @return {Integer}
def first_missing_positive(nums)
    for i in 0...nums.size
        while nums[i] - 1 != i && 
              nums[i].between?(1, nums.size) &&
              nums[nums[i] - 1] != nums[i]
              
            remote = nums[nums[i] - 1]
            nums[nums[i] - 1] = nums[i]
            nums[i] = remote
        end
    end
    puts "#{nums}"
    (0.upto(nums.size - 1).find { |i| nums[i] != i + 1 } || nums.size) + 1
end
